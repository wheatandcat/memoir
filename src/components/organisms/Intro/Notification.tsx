import React, { memo, useState } from 'react';
import { StyleSheet } from 'react-native';
import { StatusBar } from 'expo-status-bar';
import { SafeAreaView } from 'react-native-safe-area-context';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import theme from 'config/theme';
import Input from 'components/organisms/Setting/Memoir/Input';
import NotificationToggle from 'components/organisms/Setting/Memoir/Notification';
import { ConnectedType } from 'components/pages/Intro/Intro/Connected';
import Button from 'components/atoms/Button';
import dayjs from 'lib/dayjs';

export type Props = ConnectedType & {
  dayOfWeek: number;
  hours: number;
  minutes: number;
  notification: boolean;
  onNext: () => void;
};

const Notification: React.FC<Props> = (props) => {
  const [dayOfWeek, setDayOfWeek] = useState(props.dayOfWeek);
  const [time, setTime] = useState(
    dayjs(
      `0000-01-01T${('00' + props.hours).slice(-2)}:${(
        '00' + props.minutes
      ).slice(-2)}:00`
    ).toDate()
  );
  const [push, setPush] = useState(props.notification ? 1 : 0);

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar backgroundColor={theme().color.primary.main} style="dark" />
      <View style={styles.header}>
        <Text textAlign="center">初期設定</Text>
      </View>
      <View style={styles.root}>
        <View style={styles.inner}>
          <View>
            <View>
              <Text textAlign="center">いつ、ふりかえりますか？</Text>
            </View>
            <Input
              dayOfWeek={dayOfWeek}
              time={time}
              onChangeDayOfWeek={setDayOfWeek}
              onChangeTime={setTime}
            />
          </View>
          <NotificationToggle push={push} setPush={setPush} />
          <View p={3} style={styles.action}>
            <Button
              title="次へ"
              onPress={() =>
                props.onSaveNotification(
                  {
                    dayOfWeek,
                    hours: dayjs(time).hour(),
                    minutes: dayjs(time).minute(),
                    notification: push ? true : false,
                  },
                  props.onNext
                )
              }
              width={200}
            />
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    backgroundColor: theme().color.primary.main,
  },
  root: {
    backgroundColor: theme().color.background.main,
    height: '100%',
    alignItems: 'center',
  },
  header: {
    backgroundColor: theme().color.primary.main,
    height: 50,
    alignItems: 'center',
    justifyContent: 'center',
  },
  inner: {
    paddingVertical: theme().space(3),
  },
  action: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: theme().space(6),
    marginBottom: theme().space(3),
  },
});

export default memo(Notification);
