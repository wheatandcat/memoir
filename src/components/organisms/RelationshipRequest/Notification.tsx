import React, { memo } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import Image from 'components/atoms/Image';
import theme from 'config/theme';
import { MaterialIcons } from '@expo/vector-icons';

export type Props = {
  count: number;
  onPress: () => void;
};

const Notification: React.FC<Props> = (props) => {
  return (
    <TouchableOpacity onPress={props.onPress}>
      <View style={styles.root}>
        <View>
          <Image
            source={require('../../../img/icon/icon_edit.png')}
            width={35}
            height={35}
            contentFit="contain"
          />
        </View>
        <View>
          <Text size="sm">招待申請が{props.count}件届いています</Text>
        </View>
        <View>
          <MaterialIcons name="chevron-right" size={30} />
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  root: {
    backgroundColor: theme().color.primary.light,
    width: '100%',
    height: 60,
    alignItems: 'center',
    justifyContent: 'space-between',
    flexDirection: 'row',
    paddingHorizontal: theme().space(2),
  },
});

export default memo(Notification);
