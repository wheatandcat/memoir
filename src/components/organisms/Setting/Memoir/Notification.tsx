import React, { memo } from 'react';
import { StyleSheet } from 'react-native';
import SwitchSelector from 'react-native-switch-selector';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import theme from 'config/theme';

export type Props = {
  push: number;
  setPush: (push: number) => void;
};

const options = [
  { label: 'OFF', value: 0 },
  { label: 'ON', value: 1 },
];

const Notification: React.FC<Props> = (props) => {
  return (
    <>
      <View style={styles.title}>
        <View style={styles.line} />

        <View>
          <Text>通知設定</Text>
        </View>

        <View style={styles.line} />
      </View>
      <View style={styles.push}>
        <View>
          <Text variants="small">プッシュ通知</Text>
        </View>
        <View>
          <SwitchSelector
            initial={props.push}
            options={options}
            bold
            buttonColor={theme().color.primary.main}
            backgroundColor={theme().color.base.main}
            fontSize={theme().fontSizes.base}
            textColor={theme().color.base.light}
            selectedColor={theme().color.secondary.main}
            style={styles.pushInput}
            height={43}
            onPress={(v) => props.setPush(Number(v))}
          />
        </View>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  title: {
    marginTop: theme().space(5),
    flexDirection: 'row',
    height: 50,
    alignItems: 'center',
    justifyContent: 'space-around',
  },
  line: {
    width: '30%',
    backgroundColor: theme().color.base.dark,
    height: StyleSheet.hairlineWidth,
    marginHorizontal: theme().space(3),
  },
  push: {
    marginTop: theme().space(4),
    paddingHorizontal: theme().space(4),
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  pushInput: {
    width: 139,
  },
});

export default memo(Notification);
