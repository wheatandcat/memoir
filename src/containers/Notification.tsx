import * as Device from 'expo-device';
import * as Notifications from 'expo-notifications';
import React, {
  memo,
  useRef,
  useContext,
  useCallback,
  createContext,
} from 'react';
import { Platform } from 'react-native';
import {
  useCreatePushTokenMutation,
  CreatePushTokenMutationVariables,
} from 'queries/api/index';
import theme from 'config/theme';

const Context = createContext<ContextProps>({});
const { Provider } = Context;

type ContextProps = Partial<{
  onPermissionRequest: (callback: () => void) => Promise<boolean>;
}>;

type Props = {};

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

const Notification: React.FC<Props> = memo((props) => {
  const requestCallback = useRef<() => void>(() => {});

  const [createPushTokenMutation] = useCreatePushTokenMutation({
    onCompleted() {
      requestCallback.current();
    },
  });

  const onPermissionRequest = useCallback(
    async (callback: () => void) => {
      requestCallback.current = callback;

      let token;

      if (!Device.isDevice) {
        //Alert.alert('端末から実行してくだださい');
        requestCallback.current();
        return true;
      }

      const { status: existingStatus } =
        await Notifications.getPermissionsAsync();
      let finalStatus = existingStatus;

      if (existingStatus !== 'granted') {
        const { status } = await Notifications.requestPermissionsAsync();
        finalStatus = status;
      }

      if (finalStatus !== 'granted') {
        requestCallback.current();
        return false;
      }

      token = (await Notifications.getExpoPushTokenAsync()).data;
      console.log(token);

      if (Platform.OS === 'android') {
        Notifications.setNotificationChannelAsync('default', {
          name: 'default',
          importance: Notifications.AndroidImportance.MAX,
          vibrationPattern: [0, 250, 250, 250],
          lightColor: theme().color.primary.main,
        });
      }

      const variables: CreatePushTokenMutationVariables = {
        input: {
          token,
          deviceID: `${(Device.modelName || '').split(' ').join('')}-${String(
            Device.osInternalBuildId || ''
          )
            .split(' ')
            .join('')}`,
        },
      };

      createPushTokenMutation({
        variables,
      });

      return true;
    },
    [createPushTokenMutation]
  );

  return <Provider value={{ onPermissionRequest }}>{props.children}</Provider>;
});

export default Notification;

export const useNotification = () => useContext(Context);
