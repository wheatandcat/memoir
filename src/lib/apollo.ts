import { setContext } from '@apollo/client/link/context';
import { Alert } from 'react-native';
import { onError } from '@apollo/client/link/error';
import { ApolloClient, InMemoryCache, createHttpLink } from '@apollo/client';
import { storageKey, getItem } from 'lib/storage';
import Auth from 'lib/auth';

export const cache = new InMemoryCache();
const auth = new Auth();

type Param = {
  Authorization?: string;
  UserID?: string;
};

const makeApolloClient = async () => {
  const uri = `${process.env.API_HOST}/query`;

  const authLink = setContext(async (_, { headers }) => {
    const h = headers;
    const param: Param = {};
    const token = await auth.getIdToken();

    if (token) {
      param.Authorization = `Bearer ${token}`;
    } else {
      const userID = await getItem(storageKey.USER_ID_KEY);

      if (userID) {
        param.UserID = userID;
      }
    }

    return {
      headers: {
        ...h,
        ...param,
      },
    };
  });

  const errorLink = onError((error) => {
    console.log(error.graphQLErrors);
    if ((error.graphQLErrors || []).length > 0) {
      const graphQLErrors = error.graphQLErrors || [
        { message: 'エラー発生しました' },
      ];

      const message = graphQLErrors[0].message;
      Alert.alert('エラー', message);
    }
  });

  return new ApolloClient({
    link: errorLink.concat(authLink.concat(createHttpLink({ uri }))),
    cache: cache,
  });
};

export default makeApolloClient;
