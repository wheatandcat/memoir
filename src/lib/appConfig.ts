import firebase from 'lib/system/firebase';

export type AppConfig = {
  maintenance: boolean;
  maintenanceMessage: string;
  maintenancePeriod: string | null;
  supportVersion: string;
};

export const defaultAppConfig = () => ({
  maintenance: false,
  maintenanceMessage: '',
  maintenancePeriod: null,
  supportVersion: '1.0.0',
});

export const getAppConfig = async (
  db: firebase.firestore.Firestore
): Promise<AppConfig> => {
  const qs = await db.collection('appConfig').limit(1).get();

  const records = qs.docs.map((elem) => {
    return elem.data();
  });

  if (!records || !records[0]) {
    return defaultAppConfig();
  }

  return records[0] as AppConfig;
};
