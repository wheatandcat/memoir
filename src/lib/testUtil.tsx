import React from 'react';
import { GraphQLHandler, GraphQLRequest } from 'msw';
import {
  ApolloClient,
  ApolloProvider,
  createHttpLink,
  InMemoryCache,
} from '@apollo/client';
import { render } from '@testing-library/react-native';
import { server } from 'mocks/server';
import fetch from 'cross-fetch';

const link = createHttpLink({
  uri: 'http://localhost:8080/query',
  fetch,
  credentials: 'same-origin',
});

const client = new ApolloClient({
  link,
  uri: 'http://localhost:8080/query',
  cache: new InMemoryCache(),
});

export const testRenderer =
  (children: React.ReactNode) =>
  (responseOverride?: GraphQLHandler<GraphQLRequest<never>>) => {
    if (responseOverride) {
      server.use(responseOverride);
    }
    return render(<ApolloProvider client={client}>{children}</ApolloProvider>);
  };
