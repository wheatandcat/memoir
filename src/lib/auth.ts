import firebase from 'lib/system/firebase';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { storageKey } from 'lib/storage';
import dayjs from 'lib/dayjs';

const initFirebaseAuth = (): Promise<firebase.User | null> => {
  return new Promise((resolve) => {
    var unsubscribe = firebase.auth().onAuthStateChanged((user) => {
      // user オブジェクトを resolve
      resolve(user);

      // 登録解除
      unsubscribe();
    });
  });
};

class Auth {
  setSession = async (refresh = false) => {
    const user = await initFirebaseAuth();
    if (!user) {
      return null;
    }

    await AsyncStorage.setItem(storageKey.AUTH_UID_KEY, user.uid);

    const result = await user.getIdTokenResult(refresh);

    await AsyncStorage.setItem(storageKey.AUTH_ID_TOKEN_KEY, result.token);
    await AsyncStorage.setItem(
      storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY,
      String(result.claims.exp)
    );

    return result.token;
  };
  getIdToken = async () => {
    const idToken = await AsyncStorage.getItem(storageKey.AUTH_ID_TOKEN_KEY);
    if (!idToken) {
      return null;
    }

    const expiration = await AsyncStorage.getItem(
      storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY
    );

    if (Number(expiration) > dayjs().unix()) {
      return idToken;
    }

    return this.setSession(true);
  };
  logout = async () => {
    await firebase.auth().signOut();

    await AsyncStorage.removeItem(storageKey.AUTH_UID_KEY);
    await AsyncStorage.removeItem(storageKey.AUTH_ID_TOKEN_KEY);
    await AsyncStorage.removeItem(storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY);
  };
}

export default Auth;
