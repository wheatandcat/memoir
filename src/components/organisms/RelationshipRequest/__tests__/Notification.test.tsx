import React from 'react';
import { testRenderer } from 'lib/testUtil';
import { screen } from '@testing-library/react-native';
import Notification, { Props } from '../Notification';

const propsData = (): Props => ({
  count: 3,
  onPress: jest.fn(),
});

describe('components/organisms/RelationshipRequest/Notification.tsx', () => {
  it('正常にrenderすること', () => {
    testRenderer(<Notification {...propsData()} />)();
    expect(screen.findAllByText('')).toBeTruthy();
  });
});
