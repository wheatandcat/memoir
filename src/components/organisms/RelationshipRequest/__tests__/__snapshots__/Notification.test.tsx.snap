// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/RelationshipRequest/Notification.tsx 正常にrenderすること 1`] = `
<ForwardRef
  onPress={[MockFunction]}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F9E281",
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <View>
      <Image
        height={35}
        resizeMode="contain"
        source={null}
        width={35}
      />
    </View>
    <View>
      <Text
        fontFamily="RobotoCondensed-Bold"
        size="sm"
        variants="body"
      >
        招待申請が
        3
        件届いています
      </Text>
    </View>
    <View>
      <Icon
        allowFontScaling={false}
        name="chevron-right"
        size={30}
      />
    </View>
  </View>
</ForwardRef>
`;
